File: iOS/TUILiveKit/Source/State/Operation/RoomState.swift
//
//  RoomState.swift
//  VoiceRoom
//
//  Created by aby on 2024/3/7.
//

// PassthroughSubject
// CurrentValueSubject

import Foundation
import RTCRoomEngine

struct RoomState: Codable {
    var roomId: String = ""
    var createTime: UInt = 0
    var ownerInfo: User = User()
    var roomName: String = ""
    var coverURL: String = String.randomBackgroundImageUrl()
    
    
    var seatMode: TUISeatMode = .applyToTake
    var userCount: Int = 0
    var maxSeatCount: Int = 0
    var liveExtraInfo: LiveExtraInfo = LiveExtraInfo()
    
    struct LiveExtraInfo: Codable {
        var category: LiveStreamCategory = .chat
        var liveMode: LiveStreamPrivacyStatus = .public
        var maxAudienceCount: Int = 0
        var messageCount: Int = 0
        var giftIncome: Int = 0
        var giftPeopleSet: Set<String> = []
        var likeCount: Int = 0
    }
}

extension RoomState: Identifiable {
    var id: String {
        self.roomId
    }
}

extension TUISeatMode: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(UInt.self)
        self = TUISeatMode(rawValue: rawValue) ?? .applyToTake
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(self.rawValue)
    }
}

fileprivate extension String {
    static func randomBackgroundImageUrl() -> String {
        let random = arc4random() % 12 + 1
        return "https://liteav-test-1252463788.cos.ap-guangzhou.myqcloud.com/voice_room/voice_room_cover\(random).png"
    }
}

File: Flutter/livekit/lib/voice_room/state/room_state.dart
import 'package:flutter/cupertino.dart';
import 'package:tencent_live_uikit/voice_room/index.dart';
import 'package:rtc_room_engine/rtc_room_engine.dart';

class RoomState {
  String roomId = '';
  int createTime = 0;
  User ownerInfo = User();
  ValueNotifier<String> roomName = ValueNotifier('');
  ValueNotifier<String> coverUrl = ValueNotifier('https://liteav-test-1252463788.cos.ap-guangzhou.myqcloud.com/voice_room/voice_room_cover1.png');
  ValueNotifier<String> backgroundUrl = ValueNotifier('https://liteav-test-1252463788.cos.ap-guangzhou.myqcloud.com/voice_room/voice_room_background1.png');
  ValueNotifier<TUISeatMode> seatMode = ValueNotifier(TUISeatMode.applyToTake);
  ValueNotifier<int> userCount = ValueNotifier(0);
  ValueNotifier<int> maxSeatCount = ValueNotifier(0);
  ValueNotifier<LiveExtraInfo> liveExtraInfo = ValueNotifier(LiveExtraInfo());
  ValueNotifier<bool> exitRoom = ValueNotifier(false);
}

class LiveExtraInfo {
  final ValueNotifier<PrivacyStatus> liveMode = ValueNotifier(PrivacyStatus.publicity);
  int maxAudienceCount = 0;
  int messageCount = 0;
  int giftIncome = 0;
  int giftSenderCount = 0;
  int likeCount = 0;
}
